README.md
# bedrock-zcap-storage-http
Bedrock HTTP API for Authorization Capability storage


CHANGELOG.md
# bedrock-zcap-storage-http ChangeLog

## 1.3.0 - 2021-01-12

### Changed
- Update bedrock-account@5.0.

## 1.2.0 - 2020-07-07

### Changed
- Update peer deps, test deps and CI workflow.

## 1.1.0 - 2020-06-30

### Changed
- Improve test coverage.
- Update test deps.
- Update CI workflow.

## 1.0.0 - 2020-02-26

### Added
- Added core files.

- See git history for changes.


test/test.js
/*!
 * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
 */
const bedrock = require('bedrock');
require('bedrock-mongodb');
require('bedrock-zcap-storage-http');
require('bedrock-https-agent');
require('bedrock-security-context');

require('bedrock-test');
bedrock.start();


test/test.config.js
/*!
 * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const {config} = require('bedrock');
const path = require('path');

config.mocha.tests.push(path.join(__dirname, 'mocha'));

// mongodb config
config.mongodb.name = 'bedrock_zcap_storage_http_test';
config.mongodb.host = 'localhost';
config.mongodb.port = 27017;
// drop all collections on initialization
config.mongodb.dropCollections = {};
config.mongodb.dropCollections.onInit = true;
config.mongodb.dropCollections.collections = [];

// allow self-signed certs in test framework
config['https-agent'].rejectUnauthorized = false;


test/mocha/10-zcap.js
/*!
 * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const {config, util: {uuid}} = require('bedrock');
const {httpsAgent} = require('bedrock-https-agent');
const {create} = require('apisauce');
const {zcap, actors} = require('./mock-data');
const {stubPassport} = require('./helpers');

const cfg = config['zcap-storage-http'];
const {routes} = cfg;

const api = create({
  baseURL: `${config.server.baseUri}${routes.capabilities}`,
  httpsAgent,
  timeout: 1000
});

const actor = actors.alpha;

describe('capabilties API', function() {
  let capability, stubs = null;
  before(function() {
    stubs = stubPassport({actor});
  });
  beforeEach(function() {
    capability = zcap();
  });
  after(function() {
    stubs.optionallyAuthenticated.restore();
    stubs.ensureAuthenticated.restore();
  });
  describe('insert API', function() {
    it('should insert a capability', async function() {
      capability.controller = actor.id;
      const result = await api.post('', {controller: actor.id, capability});
      should.exist(result);
      result.status.should.equal(204);
    });
    it('should not insert a capability if the actor is not the controller',
      async function() {
        const result = await api.post(
          '', {controller: 'did:key:non-controller', capability});
        should.exist(result);
        result.status.should.equal(400);
      });
  });
  describe('get API', function() {
    it('should get a zcap by id', async function() {
      capability.controller = actor.id;
      const id = `urn:zcap:${uuid()}`;
      capability.id = id;
      capability.referenceId = id;
      const insertResult = await api.post(
        '', {controller: actor.id, capability});
      should.exist(insertResult);
      insertResult.status.should.equal(204);
      const getResult = await api.get('', {controller: actor.id, id});
      should.exist(getResult);
      getResult.status.should.equal(200);
      getResult.data.should.deep.equal(capability);
    });
    it('should get a zcap by referenceId', async function() {
      capability.controller = actor.id;
      const id = `urn:zcap:${uuid()}`;
      const referenceId = 'test-ref-id';
      capability.id = id;
      capability.referenceId = referenceId;
      const insertResult = await api.post(
        '', {controller: actor.id, capability});
      should.exist(insertResult);
      insertResult.status.should.equal(204);
      const getResult = await api.get('', {controller: actor.id, referenceId});
      should.exist(getResult);
      getResult.status.should.equal(200);
      getResult.data.should.deep.equal(capability);
    });
  });
  describe('delete API', function() {
    it('should delete a zcap by id', async function() {
      capability.controller = actor.id;
      const id = `urn:zcap:${uuid()}`;
      capability.id = id;
      capability.referenceId = id;
      const insertResult = await api.post(
        '', {controller: actor.id, capability});
      should.exist(insertResult);
      insertResult.status.should.equal(204);
      const deleteResult = await api.delete('', {controller: actor.id, id});
      should.exist(deleteResult);
      deleteResult.status.should.equal(204);
      const getResult = await api.get('', {controller: actor.id, id});
      should.exist(getResult);
      getResult.status.should.equal(404);
    });
    it('should delete a zcap by referenceId', async function() {
      capability.controller = actor.id;
      const id = `urn:zcap:${uuid()}`;
      const referenceId = 'test-delete-ref-id';
      capability.id = id;
      capability.referenceId = referenceId;
      const insertResult = await api.post(
        '', {controller: actor.id, capability});
      should.exist(insertResult);
      insertResult.status.should.equal(204);
      const deleteResult = await api.delete(
        '', {controller: actor.id, referenceId});
      deleteResult.status.should.equal(204);
      const getResult = await api.get('', {controller: actor.id, referenceId});
      should.exist(getResult);
      getResult.status.should.equal(404);
    });
  });
});


lib/http-zcaps.js
/*!
 * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const {asyncHandler} = require('bedrock-express');
const bedrock = require('bedrock');
const {config, util: {BedrockError}} = bedrock;
const brPassport = require('bedrock-passport');
const brZCapStorage = require('bedrock-zcap-storage');
const database = require('bedrock-mongodb');
const {ensureAuthenticated} = brPassport;

bedrock.events.on('bedrock-express.configure.routes', app => {
  const cfg = config['zcap-storage-http'];
  const {routes} = cfg;

  // insert a zcap
  app.post(
    routes.capabilities,
    ensureAuthenticated,
    // TODO: add zcap validator
    //validate('bedrock-zcap-storage-http.zcap'),
    asyncHandler(async (req, res) => {
      // TODO: consider invoking the `account` as a zcap here ... and add
      // `invoker` and `delegator` to the account to avoid having a permission
      // based model here; automatically add an admin account as an invoker
      // whenever the account is retrieved that can't be removed
      const {actor} = req.user;
      const {controller, capability} = req.body;
      const referenceId = req.body.referenceId || capability.referenceId;
      if(actor.id !== controller) {
        throw new BedrockError(
          'Permission denied.',
          'NotAllowedError', {httpStatusCode: 400, public: true});
      }
      await brZCapStorage.zcaps.insert({controller, referenceId, capability});
      res.status(204).end();
    }));

  // get one or more zcaps
  app.get(
    routes.capabilities,
    ensureAuthenticated,
    // TODO: validate query
    asyncHandler(async (req, res) => {
      // TODO: consider invoking the `account` as a zcap here ... and add
      // `invoker` and `delegator` to the account to avoid having a permission
      // based model here; automatically add an admin account as an invoker
      // whenever the account is retrieved that can't be removed
      const {actor} = req.user;
      const {controller, id, referenceId} = req.query;
      if(actor.id !== controller) {
        throw new BedrockError(
          'Permission denied.',
          'NotAllowedError', {httpStatusCode: 400, public: true});
      }
      if(id || referenceId) {
        const {capability} = await brZCapStorage.zcaps.get(
          {controller, id, referenceId});
        res.json(capability);
      } else {
        const query = {controller: database.hash(controller)};
        const results = await brZCapStorage.zcaps.find(
          {query, fields: {_id: 0, capability: 1}});
        res.json(results.map(r => r.capability));
      }
    }));

  // delete a zcap
  app.delete(
    routes.capabilities,
    ensureAuthenticated,
    // TODO: validate query
    asyncHandler(async (req, res) => {
      // TODO: consider invoking the `account` as a zcap here ... and add
      // `invoker` and `delegator` to the account to avoid having a permission
      // based model here; automatically add an admin account as an invoker
      // whenever the account is retrieved that can't be removed
      const {actor} = req.user;
      const {controller, id, referenceId} = req.query;
      if(actor.id !== controller) {
        throw new BedrockError(
          'Permission denied.',
          'NotAllowedError', {httpStatusCode: 400, public: true});
      }
      const removed = await brZCapStorage.zcaps.remove(
        {controller, id, referenceId});
      if(removed) {
        res.status(204).end();
      } else {
        res.status(404).end();
      }
    }));
});


lib/http.js
/*!
 * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const asyncHandler = require('express-async-handler');
const bedrock = require('bedrock');
const brPassport = require('bedrock-passport');
const brZCapStorage = require('bedrock-zcap-storage');
const {config} = bedrock;
const database = require('bedrock-mongodb');
require('bedrock-express');
const {
  ensureAuthenticated
} = brPassport;
const {util: {BedrockError}} = bedrock;

// load config defaults
require('./config');

// module API
const api = {};
module.exports = api;

bedrock.events.on('bedrock-express.configure.routes', app => {
  const cfg = config['zcap-storage-http'];
  const {routes} = cfg;

  // insert a zcap
  app.post(
    routes.zcaps,
    ensureAuthenticated,
    // TODO: add zcap validator
    //validate('bedrock-zcap-storage-http.zcap'),
    asyncHandler(async (req, res) => {
      // TODO: consider invoking the `account` as a zcap here ... and add
      // `invoker` and `delegator` to the account to avoid having a permission
      // based model here; automatically add an admin account as an invoker
      // whenever the account is retrieved that can't be removed
      const {actor} = req.user;
      const {controller, capability} = req.body;
      const referenceId = req.body.referenceId || capability.referenceId;
      if(actor.id !== controller) {
        throw new BedrockError(
          'Permission denied.',
          'NotAllowedError', {httpStatusCode: 400, public: true});
      }
      await brZCapStorage.zcaps.insert({controller, referenceId, capability});
      res.status(204).end();
    }));

  // get one or more zcaps
  app.get(
    routes.zcaps,
    ensureAuthenticated,
    // TODO: validate query
    asyncHandler(async (req, res) => {
      // TODO: consider invoking the `account` as a zcap here ... and add
      // `invoker` and `delegator` to the account to avoid having a permission
      // based model here; automatically add an admin account as an invoker
      // whenever the account is retrieved that can't be removed
      const {actor} = req.user;
      const {controller, id, referenceId} = req.query;
      if(actor.id !== controller) {
        throw new BedrockError(
          'Permission denied.',
          'NotAllowedError', {httpStatusCode: 400, public: true});
      }
      if(id || referenceId) {
        const {capability} = await brZCapStorage.zcaps.get(
          {controller, id, referenceId});
        res.json(capability);
      } else {
        const query = {controller: database.hash(controller)};
        const results = await brZCapStorage.zcaps.find(
          {query, fields: {_id: 0, capability: 1}});
        res.json(results.map(r => r.capability));
      }
    }));

  // delete a zcap
  app.delete(
    routes.zcaps,
    ensureAuthenticated,
    // TODO: validate query
    asyncHandler(async (req, res) => {
      // TODO: consider invoking the `account` as a zcap here ... and add
      // `invoker` and `delegator` to the account to avoid having a permission
      // based model here; automatically add an admin account as an invoker
      // whenever the account is retrieved that can't be removed
      const {actor} = req.user;
      const {controller, id, referenceId} = req.query;
      if(actor.id !== controller) {
        throw new BedrockError(
          'Permission denied.',
          'NotAllowedError', {httpStatusCode: 400, public: true});
      }
      const removed = await brZCapStorage.zcaps.remove(
        {controller, id, referenceId});
      if(removed) {
        res.status(204).end();
      } else {
        res.status(404).end();
      }
    }));
});
